ios/engine/KMEI/KeymanEngineTests/ResourceUpdateTests.swift

```diff
@@ -240,47 +240,256 @@ class ResourceUpdateTests: XCTestCase {
+    let updateExpectation = XCTestExpectation(description: "performBatchUpdate should return cleanly")
+
+    let updateSet = downloadManager.getKeysForUpdatablePackages()
+    downloadManager.performBatchUpdate(forPackageKeys: updateSet, withNotifications: false) { successes, failures in
+      XCTAssertEqual(failures.count, 0)
+      XCTAssertEqual(successes.count, updateSet.count)
+
+      updateExpectation.fulfill()
+    }
+
+    // Step 2.5:  ensure all mocking is properly established.
+
+    // At this stage, the downloading queue is paused.  We need to inspect the batch ordering in order
+    // to get the mocking right.
+    //
+    // Swift implements 'secure hashing' behavior, which means the order is technically
+    // inconsistent between runs.  (In order to prevent "hash flooding" attacks.)
+    // Reference: https://www.hackingwithswift.com/articles/115/swift-4-2-improves-hashable-with-a-new-hasher-struct
+    guard case let .compositeBatch(updateBatch) = downloadManager.downloader.topLevelNodes()[0] else {
+      XCTFail()
+      return
+    }
+
+    // Now that we are examining the batch node, we can determine the order in which KeymanEngine wants to
+    // download the updates and mock accordingly.
+    updateBatch.batchQueue.forEach { entry in
+      // Create the mock corresponding to this key and add it to the mocking queue.
+      let packageLocation = TestUtils.Packages.getURLForPackage(withKey: entry.0.packageKeys[0])!
+      let batchForKey = TestUtils.Downloading.MockResult(location: packageLocation, error: nil)
+      mockedURLSession.queueMockResult(.download(batchForKey))
+    }
+
+    // Step 3:  Let the updater do its thing.
+
+    downloadManager.downloader.autoExecute = true // We'll let it auto-run for this test.
+    downloadManager.downloader.step() // It will auto-run after this.
+
+    wait(for: [updateExpectation], timeout: 15)
+
+    // Step 4:  Assertion time.
+
+    let packages = ResourceFileManager.shared.installedPackages
+    packages.forEach { package in
+      // Baseline - the kmp.json files should no longer be the old, auto-generated versions.
+      XCTAssertFalse(package.metadata.isAutogeneratedWrapper, "\(package.key.id) @ \(package.sourceFolder)")
+      XCTAssertNotEqual(package.version, Version.freshInstall)
+
+      // Most of the packages have author information included within their metadata.
+      // This is never autogenerated.
+      if package.key != TestUtils.Packages.Keys.indigenous_nt &&
+         package.key != TestUtils.Packages.Keys.sil_euro_latin {
+        XCTAssertNotNil(package.metadata.info?.author, package.key.id)
+      }
+
+      if let folderContents = try? FileManager.default.contentsOfDirectory(atPath: package.sourceFolder.path) {
+        if package.resourceType() == .keyboard {
+          // There is a 'desktop' keyboard file within the distributed package.
+          XCTAssertTrue(folderContents.contains(where: { $0.hasSuffix(".kmx") }))
```

Comment:
Technically, this does not necessarily have to be true; it just happens to be true for the packages you are testing. Might be worth noting that for future readers.